incubator/binding-mqtt-kafka/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/kafka/internal/stream/MqttKafkaPublishFactory.java

```diff
@@ -280,7 +266,36 @@ private void onMqttBegin(
 
             assert initialAck <= initialSeq;
 
-            delegate.doKafkaBegin(traceId, authorization, affinity);
+            final OctetsFW extension = begin.extension();
+            final MqttBeginExFW mqttBeginEx = extension.get(mqttBeginExRO::tryWrap);
+
+            assert mqttBeginEx.kind() == MqttBeginExFW.KIND_PUBLISH;
+            final MqttPublishBeginExFW mqttPublishBeginEx = mqttBeginEx.publish();
+            String topicName = mqttPublishBeginEx.topic().asString();
+            assert topicName != null;
+
+            String[] topicHeaders = topicName.split("/");
+            topicNameHeaders = new OctetsFW[topicHeaders.length];
+            for (int i = 0; i < topicHeaders.length; i++)
+            {
+                String16FW topicHeader = new String16FW(topicHeaders[i]);
+                topicNameHeaders[i] = new OctetsFW().wrap(topicHeader.value(), 0, topicHeader.length());
+            }
+            clientIdOctets = new OctetsFW()
+                .wrap(mqttPublishBeginEx.clientId().value(), 0, mqttPublishBeginEx.clientId().length());
+            final DirectBuffer topicNameBuffer = mqttPublishBeginEx.topic().value();
+            key = new KafkaKeyFW.Builder()
+                .wrap(keyBuffer, 0, keyBuffer.capacity())
+                .length(topicNameBuffer.capacity())
+                .value(topicNameBuffer, 0, topicNameBuffer.capacity())
+                .build();
```

Comment:
The storage for this `key`, i.e. the `keyBuffer`, is shared and will be overwritten by subsequent publish streams for different mqtt topics on the same core.

I think we have this backwards. The flyweight and builder can be shared, as they are simply strongly-typed access to the underlying key buffer, but the key buffer itself cannot be shared as it would be overwritten.

The same feedback applies to the client id and the topic headers.